common/cuda_hip/solver/lobpcg_kernels.cpp

```diff
@@ -0,0 +1,64 @@
+// SPDX-FileCopyrightText: 2017 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/solver/lobpcg_kernels.hpp"
+
+#include <ginkgo/core/base/types.hpp>
+
+#include "common/cuda_hip/base/dev_lapack_bindings.hpp"
+
+#if GKO_HAVE_LAPACK
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+namespace lobpcg {
+
+
+using gko::kernels::lobpcg::workspace_mode;
+
+template <typename ValueType>
+void symm_generalized_eig(std::shared_ptr<const DefaultExecutor> exec,
+                          const workspace_mode alloc,
+                          matrix::Dense<ValueType>* a,
+                          matrix::Dense<ValueType>* b,
+                          array<remove_complex<ValueType>>* e_vals,
+                          array<char>* workspace)
+{
+    const auto id = exec->get_device_id();
+    auto handle = exec->get_dev_lapack_handle();
+
+    int n = static_cast<int>(a->get_size()[1]);  // column-major
+    int lda = static_cast<int>(a->get_stride());
+    int ldb = static_cast<int>(b->get_stride());
+    int fp_buffer_num_elems;
+    if (alloc == workspace_mode::allocate) {
+        dev_lapack::sygvd_buffersize(handle, LAPACK_EIG_TYPE_1,
+                                     LAPACK_EIG_VECTOR, LAPACK_FILL_LOWER, n,
+                                     a->get_values(), lda, b->get_values(), ldb,
+                                     e_vals->get_data(), &fp_buffer_num_elems);
+        size_type total_bytes = sizeof(ValueType) * fp_buffer_num_elems;
+        workspace->resize_and_reset(total_bytes);
+    } else {
+        fp_buffer_num_elems = workspace->get_size() / sizeof(ValueType);
+    }
+    array<int> dev_info(exec, 1);
+    dev_lapack::sygvd(handle, LAPACK_EIG_TYPE_1, LAPACK_EIG_VECTOR,
+                      LAPACK_FILL_LOWER, n, a->get_values(), lda,
+                      b->get_values(), ldb, e_vals->get_data(),
+                      (ValueType*)workspace->get_data(), fp_buffer_num_elems,
```

Comment:
```suggestion
                      static_cast<ValueType*>(workspace->get_data()), fp_buffer_num_elems,
```

Reply:
`static_cast` isn't allowed here, but I can do `reinterpret_cast`, unless there is a better way?

Reply:
ah, it is already char*, then yes you can not use static_cast.

Reply:
The C-style casts should be removed now.